Add VS Code Extension

Burhan-Q opened the pull request:
## Summary
Adds check to see if running in VS Code terminal. If VS Code terminal found, will additionally check if [Ultralytics-Snippets](https://marketplace.visualstudio.com/VSCode) extension is installed and if not, will show logging message. This was a recommendation from @glenn-jocher at the meeting on Monday 2024-08-05.

## Screenshots

Linux system without installed extension
![image](https://github.com/user-attachments/assets/61ed031b-e552-46b4-ba12-64e21b9c702a)

Windows system with installed extension
![image](https://github.com/user-attachments/assets/73bb8c64-3e6b-4159-b9cc-8466d398a56a)

Verified on Windows that uninstalling the extension would show prompt again.

## Notes
- I wasn't certain where the best location for the logging message would be, and open for suggestions to change
- Logging message shown at start of command or after first `ultralytics` import
- Checking for install looks for files, but these could be present even when user uninstalls extension
  - parsing extensions JSON file to check for install probably too much additional

<!--
Thank you 🙏 for your contribution to [Ultralytics](https://ultralytics.com) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. **Check for Existing Contributions**: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. **Link Related Issues**: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. **Elaborate Your Changes**: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. **Ultralytics Contributor License Agreement (CLA)**: To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    _I have read the CLA Document and I sign the CLA_

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced VS Code support with a new extension for improved development efficiency using Ultralytics.

### 📊 Key Changes
- 📜 Added documentation for **Ultralytics-Snippets** VS Code extension.
- 🆕 Introduced a new file for the **VS Code Integration**.
- 🚀 Updated settings to include a prompt to install the VS Code extension when detected.
- 🛠️ Modified existing settings to incorporate the new `vscode_msg` setting.
- ⚙️ Adjusted the initialization of settings to a new version (`0.0.5`).

### 🎯 Purpose & Impact
- **Ease of Use:** The VS Code extension provides pre-built code snippets, reducing the need to write repetitive code and helping with quick access to common functionalities.
- **Productivity Boost:** With intelligent code completion and predefined snippets, developers can write cleaner, error-free code faster, enhancing overall productivity.
- **Improved Learning Curve:** New users and experienced developers alike can benefit from examples and streamlined development workflows, making it easier to adopt and use the Ultralytics Python package.
- **Accessibility:** Adding prompts to install the extension ensures that developers who use VS Code get the recommendations needed to enhance their development environment.

Burhan-Q committed:
Adds VS Code check and recommend Ultralytics-Snippets

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q committed:
adds opt out PATH variable and checks obsolete file for uninstalled extension

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q commented:
Included check for `"VSCODE_NO_RECOMMEND"` to give users ability to opt-out of prompt without installing extension. Also checking `.obsolete` file as extension directory does not get removed on uninstall, so users will be prompted again after uninstall.

ambitious-octopus commented on ultralytics/utils/__init__.py:
```diff
@@ -1080,3 +1081,14 @@ def url2file(url):
+    if not installed:
+        LOGGER.info(
+            colorstr("VS Code terminal detected.\n")
+            + "  Enhance your Ultralytics experience by installing Ultralytics-Snippets for VS Code ⚡.\n"
```
It would be helpful to add a link here!

ambitious-octopus reviewed (commented):
Tested with vscode snap, flatpack and deb package. Works in all three cases!

glenn-jocher commented:
@Burhan-Q ok nice. Let's drop all this into a function, and we need to think about how not to show this every time to keep from annoying users. 

Maybe we could show this only on the commands that return significant console output and that the user is likely to see, i.e. train/predict/export/val end commands that point to the docs pages. We also need to provide them a way to disable and notify them about it, likely in the settings, i.e. `yolo settings vscode_msg False`.

Otherwise I can see users getting extremely frustrated seeing this on each ultralytics import.

Burhan-Q commented:
## TODOs
- [ ] Migrate code into a function
- [ ] Add a permanent settings key to enable/disable prompt, i.e. `yolo settings vscode False` to disable integration message, just like we disable the other integrations
 - [ ] Only show occasionally. 
   - Docs prompts are a good example, at the end of each mode command (i.e. at the end of the `train/predict/val` console output)
- [ ] Create a docs integration page that clearly explains how to install, uninstall, disable, enable etc. the integration, and link to that docs page directly (just like the existing message you see on `train/predict/val` etc.).

Burhan-Q committed:
add `vscode_msg` to SettingsManager and increment version

Burhan-Q committed:
Migrate VS Code extension recommendation logic to function

Burhan-Q committed:
function returns string, add docs integrations page url

Burhan-Q committed:
add VS Code extension recommendation if all conditions met

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q commented:
Now prompts after `predict` if not opted out, extension not installed, and VS Code detected.

![image](https://github.com/user-attachments/assets/697f027c-33f5-488e-a6d0-23fff75ad7bd)

Burhan-Q committed:
add vscode snippets integration documentation

Burhan-Q committed:
Add entry for VS Code integration

Burhan-Q committed:
update settings table with vscode_msg setting, plus clean up and add links for Settings table

Burhan-Q committed:
Merge branch 'is_vscode' of https://github.com/ultralytics/ultralytics into is_vscode

Burhan-Q committed:
add keys, update emoji, add to FAQ, formatting, add link, fix relative link

Burhan-Q commented:
## Images for docs integration page

<details><summary>partial snippet alias</summary>
<p>

![image](https://github.com/user-attachments/assets/bdd9aada-de8a-4d73-b714-04f8a48c0436)

</p>
</details> 

<details><summary>predict kwargs preview</summary>
<p>

![image](https://github.com/user-attachments/assets/63b17bef-f755-43f5-bc61-ad092fb6fcfb)

</p>
</details> 

<details><summary>results loop preview</summary>
<p>

![image](https://github.com/user-attachments/assets/16cc72eb-4390-4380-bfaa-a0e84e5251db)

</p>
</details> 

<details><summary>fields updates</summary>
<p>

![ultralytics-snippets_multi-update](https://github.com/user-attachments/assets/661c1b85-39a5-48ec-9faa-89d44e49aa37)

</p>
</details>

Burhan-Q committed:
add vscode-snippets page

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference Section by https://ultralytics.com/actions

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
add description, add conclusion, updates, fixes, reformats, and clean up.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q committed:
add newline for Note section

Burhan-Q commented:
I also updated `quickstart.md` table to include the new setting. While I was there, I added some links to the various integration pages, but left any existing external URLs. I also organized the links for the table at the bottom of the document for a cleaner doc. 

Other requested changes/features added. Will wait to mark as "ready" until I get some others to do a quick proof read of the new integration page.

Burhan-Q commented:
@lakshanthad and @ambitious-octopus can I get a proof read on the documentation when you have some time please 🙌

Burhan-Q committed:
Merge branch 'main' into is_vscode

ambitious-octopus reviewed (commented):
The documentation looks perfect! But there is one feature that is not described: the extension is invoked even without writing ultra. 
As in these examples

![image](https://github.com/user-attachments/assets/586075bc-a884-494c-9e1d-5e5fa8648fa6)

![image](https://github.com/user-attachments/assets/223a7ebc-5ca3-44ed-a979-a9258f339955)

![image](https://github.com/user-attachments/assets/6e4aff97-def6-4fcd-aebb-81ffcb79b03d)


Is it worth mentioning what do you think @Burhan-Q ?

Burhan-Q commented:
@ambitious-octopus I did mention this a bit. The section on snippet completion.

> It's not required to type the full prefix of the snippet, or even from the start of the snippet. The snippets are named in the most descriptive way possible, but this means there could be a lot to type and that would be counterproductive if the aim is to move _faster_. Luckily VS Code lets users type `ultra.example-yolo-predict`, `example-yolo-predict`, `yolo-predict`, or even `ex-yolo-p` and still reach the intended snippet option!

Burhan-Q commented on docs/en/integrations/vscode-snippets.md:
```diff
@@ -0,0 +1,271 @@
+
+If you use Visual Studio Code and answered 'yes' to any of the above, then the Ultralytics-snippets extension for VS Code is here to help! Read on to learn more about the extension, how to install it, and how to use it.
+
+## Inspired by the Ultralytics Community
```
There is a link to the repo in a couple later sections. I think it's better to place it further down the page.

Burhan-Q commented on docs/en/integrations/vscode-snippets.md:
```diff
@@ -0,0 +1,271 @@
+
+<p align="center">
+  <br>
+    <img src="https://github.com/Burhan-Q/ultralytics-snippets/assets/62214284/42ad0b17-e752-479c-9c6c-e451fffbe8b3" alt="Snippet Prediction Preview">
```
I see what you're saying. Here's what my thoughts were when making this; I made an assumption that many new users would be viewing:
- Opening banner is supposed to provide context about what is going to happen.
- End banner is supposed help viewers understand that the operation is completed.
- Timer is there to show how fast running the example is when inserting via the snippet.

I won't argue that they draw attention, but this _was_ intentional. If there's a lot going on in the anmimation (that's relevant), it means that viewers will watch the loop multiple times to catch everything. 

@ambitious-octopus what do you mean abou the GIF being _slow_? As in a low frame rate or that it's too long (duration)? I tried to reduce the file size as much as possible without losing _too_ much quality or information/frames.

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
address review comments and update keyboard keys to use `kpb` tag

Burhan-Q committed:
Merge branch 'is_vscode' of https://github.com/ultralytics/ultralytics into is_vscode

lakshanthad commented:
> @lakshanthad and @ambitious-octopus can I get a proof read on the documentation when you have some time please 🙌

Yes. Will do this today and share feedback @Burhan-Q

Burhan-Q committed:
Merge branch 'main' into is_vscode

lakshanthad commented:
@Burhan-Q, I have reviewed the documentation and also tried the extension. It is already well documented along with great functionality of the snippet! Excellent work overall 🔥

I just have one small feedback about the GIF in the beginning. In the GIF, you are using `ultra.example-yolo`. How about we change to `yolo-p` since it will make it look easier to type since it is also supported?

Burhan-Q commented:
@lakshanthad thanks man!

> I just have one small feedback about the GIF in the beginning. In the GIF, you are using ultra.example-yolo. How about we change to yolo-p since it will make it look easier to type?

It's possible to type only partial snippets. I tried to make the names _as obvious_ as possible, but I know that tends to make them long to type out. I mentioned in this comment https://github.com/ultralytics/ultralytics/pull/15027#issuecomment-2277505947 (with a quote from the doc) that the entire snippet prefix does not need to be used. 

Since it was brought up twice, I think that means I need to draw more attention to this. I think I'll include a "tip" admonition to make it stand out more.

lakshanthad commented:
@Burhan-Q I know it is possible to type partial snippets which I saw in your documentation. What I was pointing out was to use 'yolo-p' in the GIF to make the GIF look even simple!

Burhan-Q commented:
> @Burhan-Q I know it is possible to type partial snippets which I saw in your documentation. What I was pointing out was to use `yolo-p` in the GIF to make the GIF look even simple!

@lakshanthad I see what you mean. 🤔 I see what you're saying. I think I could do an A | B test for that in the future. For now, my preference is to keep the fully written (even tho I don't actually type all of it) prefix, so it's a bit more obvious. The input is appreciated!

Burhan-Q commented:
![image](https://github.com/user-attachments/assets/d9e74622-8818-42fb-82db-603cfe97de7c)

Burhan-Q committed:
add highlight for shortening snippet prefix

Burhan-Q committed:
Merge branch 'is_vscode' of https://github.com/ultralytics/ultralytics into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

UltralyticsAssistant committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
include comment on obsolete file purpose

Burhan-Q committed:
update conditional

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

UltralyticsAssistant committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

UltralyticsAssistant committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

Burhan-Q committed:
Merge branch 'main' into is_vscode

UltralyticsAssistant committed:
Merge branch 'main' into is_vscode

glenn-jocher committed:
Merge branch 'main' into is_vscode

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Update __init__.py

glenn-jocher committed:
Update __init__.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Update __init__.py

glenn-jocher committed:
Update mkdocs.yml

glenn-jocher committed:
Update index.md

glenn-jocher committed:
Rename vscode-snippets.md to vscode.md

glenn-jocher committed:
Update vscode.md

glenn-jocher committed:
Update __init__.py

glenn-jocher committed:
Update __init__.py

glenn-jocher committed:
Update vscode.md

glenn-jocher committed:
Update quickstart.md

glenn-jocher committed:
Update __init__.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Update __init__.py
